client/rpcserver/rpcserver.go

```diff
@@ -1,20 +1,411 @@
+	mux.Use(middleware.Recoverer)
+	mux.Use(middleware.RealIP)
+	mux.Use(s.authMiddleware)
+
+	// Websocket endpoint
+	mux.Get("/ws", s.handleWS)
+
+	// https endpoint
+	mux.Post("/", s.handleJSON)
+
+	return s, nil
+}
+
+// Run starts the web server. Satisfies the runner.Runner interface.
+func (s *RPCServer) Run(ctx context.Context) {
+	s.ctx = ctx
+	// Close the listener on context cancellation.
+	go func() {
+		<-ctx.Done()
+		err := s.listener.Close()
+		if err != nil {
+			log.Errorf("Problem shutting down rpc: %v", err)
+		}
+		s.mtx.Lock()
+		defer s.mtx.Unlock()
+		for _, cl := range s.clients {
+			cl.Disconnect()
+		}
+	}()
+	log.Infof("RPC server listening on %s", s.listener.Addr())
+	err := s.srv.Serve(s.listener)
+	if err != http.ErrServerClosed {
+		log.Warnf("unexpected (http.Server).Serve error: %v", err)
+	}
 	log.Infof("RPC server off")
 }
+
+// watchMarket watches the specified market. A fresh order book and a quit
+// function are returned on success. The quit function should be called to
+// unsubsribe the client from the market.
+func (s *RPCServer) watchMarket(cl *wsClient, dex string, base, quote uint32) (book *core.OrderBook, quit func(), err error) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	mktID := marketID(base, quote)
+	syncer, found := s.syncers[mktID]
+	if !found {
+		syncer, err = newMarketSyncer(s.ctx, s.core, dex, base, quote)
+		if err != nil {
+			return
+		}
+		s.syncers[mktID] = syncer
+	}
+	book = syncer.add(cl)
+	return book, func() {
+		syncer.remove(cl)
+	}, nil
+}
+
+// handleRequest sends the request to the correct handler function if able.
+func (s *RPCServer) handleRequest(cl *wsClient, req *msgjson.Message) *msgjson.Error {
+	if req.Route == "" {
+		log.Debugf("received empty request")
+		return msgjson.NewError(msgjson.RPCErrorUnspecified, "no route was supplied")
+	}
+
+	// Find the correct handler for this route.
+	h, ok := routes[req.Route]
+
+	// If this is a ws request, check ws routes.
+	if !ok && cl != nil {
+		h, ok = wsHandlers[req.Route]
+	}
+
+	if !ok {
+		log.Debugf("could not find route: %v", req.Route)
+		return msgjson.NewError(msgjson.RPCUnknownRoute, "unknown route")
+	}
+
+	// req will be populated with a response after this.
+	return h(s, cl, req)
+}
+
+// parseRequest parses the msgjson message in the request body and creates a
+// response message.
+func (s *RPCServer) parseHTTPRequest(r *http.Request) (msg *msgjson.Message) {
+	msg = &msgjson.Message{Type: msgjson.Response}
+	payload := new(msgjson.ResponsePayload)
+
+	// This will be filled with the msgjson request.
+	req := new(msgjson.Message)
+
+	defer func() {
+		// If an error was encountered, marshal the payload, otherwise
+		// the handler will take care of it.
+		if payload.Error != nil {
+			p, err := encode(req.Route, payload)
+			if err != nil {
+				msg.Payload = encodingFailurePayload
+			}
+			msg.Payload = p
+		} else {
+			// If no error was encountered, req has been filled with
+			// a response.
+			*msg = *req
```

Comment:
This type of pattern is why the handlers should return the response `*Message` directly. In fact, this entire function becomes a lot simpler when the handlers assemble their own responses.